### PR TITLE
Add support for fp8 -> bfp8 for tilize op

### PR DESCRIPTION
**Pull request by jvegaTT**
### PR Category
Bug fix

### Summary
This PR is to fix issue [#44114](https://github.com/tenstorrent/tt-metal/issues/44114) solving tilize op's problem on fp8 -> bfp8.

### What is changed

1. Updated test cases.
2. Guaranteed DRAM → L1 alignment during copy operations by introducing an intermediate tensor.
3. Added value padding for FP8 in the tilize operator.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:jvega/tilize_and_typecast2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:jvega/tilize_and_typecast2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:jvega/tilize_and_typecast2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:jvega/tilize_and_typecast2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:jvega/tilize_and_typecast2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:jvega/tilize_and_typecast2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:jvega/tilize_and_typecast2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:jvega/tilize_and_typecast2)

### CODE REVIEW

**Comment by jvegaTT** [comment]
This PR is blocked by https://github.com/tenstorrent/tt-metal/pull/43775

**Comment by nkapreTT** [comment]
This closes #44312 and #44114 from the Python/ziveli side cleanly. Confirmed by reading the diff:

- `tt_metal/api/tt-metalium/experimental/tensor/tensor_types.hpp:35-39` adds `DataType::FP8_E4M3 = 8`
- `ttnn/cpp/ttnn-nanobind/ttnn_dtype_traits.hpp:266+` exposes it through nanobind; will surface as `ttnn.DataType.FP8_E4M3` in Python
- `tilize_device_operation.cpp:109` accepts FP8_E4M3 input when `output_dtype=BFLOAT8_B`

On the tt-ziveli compiler side, once this merges we'll:
1. Rebase matmul-helpers branch / clear our local `_str_to_ttnn_dtype` workaround that maps `FLOAT8_E4M3FN` → `ttnn.uint8` (the deepseek_prefill convention we copied) — point it at `ttnn.DataType.FP8_E4M3` instead
2. Re-run the silicon test for our `tilize_fp8_to_bfp8` InspectCase — expecting it to land at the same PCC > 0.999 that the bf16→bfp8 sibling already hits with our cross-class composition helper

The dlpack-code=UInt workaround for torch ≤ 2.7 is fine — `view()`-based reinterpret in `to_torch` does the right thing.

Thanks for the fast turnaround on this.

File: models/demos/deepseek_v3_d_p/tests/test_prefill_block.py [hunk 1]
```diff
@@ -11,8 +11,6 @@
 extracts those weights into our TT state_dict format, and compares forward passes.
 """
 
-from pathlib import Path
-
 import pytest
```
[the remark this answers is not in the document]
Comment (ronnie00TT) [reply models/demos/deepseek_v3_d_p/tests/test_prefill_block.py:16]: Restored

**Review by riverwuTT — APPROVED** [review]
LGTM on tensor related changes.

File: ttnn/cpp/ttnn/operations/data_movement/tilize/device/tilize_multi_core_block_program_factory.cpp [hunk 2]
```diff
@@ -29,6 +29,16 @@ void push_cb_pair(
     uint32_t num_tiles,
     tt::DataFormat input_cb_data_format,
     tt::DataFormat output_cb_data_format) {
+    // 64 is added to solve the mis-alignment problem when loading from DRAM to L1
+    desc.cbs.push_back(CBDescriptor{
+        .total_size = input_single_tile_size + 64,
```
Comment (aliaksei-sala) [inline ttnn/cpp/ttnn/operations/data_movement/tilize/device/tilize_multi_core_block_program_factory.cpp]: We can get alignment from metal instead of hardcoding the value
`uint32_t dram_alignment = tt::tt_metal::hal::get_dram_alignment();`
and use round_up to that value instead, wdyt ?

File: models/demos/deepseek_v3/tests/unit/test_tilize.py [hunk 3]
```diff
@@ -84,3 +83,53 @@ def check_op(tt_output):
         assert_with_pcc(torch_output, tt_output, 0.9999)
 
     run_test(mesh_device, run_op, check_op, enable_trace)
+
+
+@pytest.mark.parametrize("mesh_device", [(8, 8)], indirect=True)
```
Comment (aliaksei-sala) [inline models/demos/deepseek_v3/tests/unit/test_tilize.py:88]: If we use smaller shapes, operation_attributes.[enough_space_height](https://github.com/tenstorrent/tt-metal/blob/28cd5269aee34206f62de610b047ae52164b340d/ttnn/cpp/ttnn/operations/data_movement/tilize/device/tilize_device_operation.cpp#L192) will be true and op will select TilizeMultiCoreDefaultProgramFactory

Could you please add `fp8` support for that program factory as well ?

**Review by aliaksei-sala — APPROVED** [review]

File: models/demos/deepseek_v3/tests/unit/test_tilize.py (hunk 3 above)
Comment (ronnie00TT) [reply models/demos/deepseek_v3/tests/unit/test_tilize.py:88]: I think TilizeMultiCoreDefaultProgramFactory has already supported `fp8`. Since I've tested smaller shape (1, 1, 800, 1024) locally. It called TilizeMultiCoreDefaultProgramFactory and test passed.

File: ttnn/cpp/ttnn/operations/data_movement/tilize/device/tilize_multi_core_block_program_factory.cpp (hunk 2 above)
Comment (ronnie00TT) [reply ttnn/cpp/ttnn/operations/data_movement/tilize/device/tilize_multi_core_block_program_factory.cpp]: Thanks for pointing out that. Yes, changing the magic number could help generalize to other architectures.
I've already done that. Also, I've changed other magic numbers in [commit](https://github.com/tenstorrent/tt-metal/pull/44307/changes/8d76a87003dd727592080f2e05e84ee2cd81f78b).

**Review by yalrawwashTT — APPROVED** [review]

**Review by yieldthought — APPROVED** [review]
🔥

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/device/kernels/dataflow/reader_unary_pad_multicore_both_dims.cpp [hunk 4]
```diff
@@ -74,19 +105,47 @@ void kernel_main() {
         uint32_t original_addr = get_write_ptr(cb_id_in0);
         for (uint32_t k = start_row_id; k < start_row_id + num_rows; k++) {
             uint64_t src_noc_addr = s.get_noc_addr(size_2d + k);
+            if (((src_noc_addr + (uint64_t)start_column_id) & dram_align_offset) ==
+                ((uint64_t)l1_write_addr & dram_align_offset)) {
+                // Read from DRAM to tmp buffer
+                noc_async_read(src_noc_addr + (uint64_t)start_column_id, (uint64_t)l1_write_addr, width_size);
```
Comment (amorrisonTT) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/device/kernels/dataflow/reader_unary_pad_multicore_both_dims.cpp]: Nit:
```suggestion
                noc_async_read(src_noc_addr + (uint64_t)start_column_id, l1_write_addr, width_size);
```
And other examples

File: ttnn/cpp/ttnn/operations/copy/typecast/typecast.cpp [hunk 5]
```diff
@@ -38,7 +38,8 @@ inline Tensor typecast_impl(
         (input_dtype == DataType::UINT8 and output_dtype != DataType::BFLOAT16);
     bool fp32_dest_acc_en = preserve_fp32_precision or output_dtype == DataType::UINT32 or
                             output_dtype == DataType::INT32 or output_dtype == DataType::FLOAT32 or
-                            input_dtype == DataType::UINT32 or input_dtype == DataType::INT32;
+                            input_dtype == DataType::UINT32 or input_dtype == DataType::INT32 or
+                            output_dtype == DataType::FP8_E4M3 or input_dtype == DataType::FP8_E4M3;
```
Comment (amorrisonTT) [inline ttnn/cpp/ttnn/operations/copy/typecast/typecast.cpp:42]: Jw, how does typecast get pulled in to this PR?

**Review by amorrisonTT — APPROVED** [review]
Approved with minor nit

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/device/kernels/dataflow/reader_unary_pad_multicore_both_dims.cpp (hunk 4 above)
Comment (ronnie00TT) [reply ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/device/kernels/dataflow/reader_unary_pad_multicore_both_dims.cpp]: Thanks! Now fixed.

File: ttnn/cpp/ttnn/operations/copy/typecast/typecast.cpp (hunk 5 above)
Comment (ronnie00TT) [reply ttnn/cpp/ttnn/operations/copy/typecast/typecast.cpp:42]: Actually, this is not required for tilize op, but it's required for the typecast [operation](https://github.com/tenstorrent/tt-metal/blob/238063231c56184e1b6f6864f36778619abc07ca/models/demos/deepseek_v3/tests/unit/test_tilize.py#L125). And `fp32_dest_acc_en` is required if the output dtype is `FP8_E4M3`

**Review by ntarafdar — APPROVED** [review]
approving to unblock, please make sure we test with unaligned shapes on both WH and BH to test the alignment.

**Comment by jvegaTT** [comment]
@ntarafdar padding is not supported on bfp8 so the shape at least on the last dim has to be tile aligned, also this data format is only available in blackhole
